### PR TITLE
Add functionality to support IA3

### PR DESCRIPTION
**Pull request by SumanthRH**
Hi,

We've added some code to support  IA3 from the [T-few paper](https://arxiv.org/abs/2205.05638). Most of the code is inspired by the LoRA implementation. Currently, implementation supports multiple adapters, int-8 training, merging and unmerging. I've only added a minimal set of models for now. With IA3, there are learned vectors added to key, value and feed-forward layers. The forward pass differs for feedforward vs non-feedforward layers. 

I hope this can be merged into the repo!

### CODE REVIEW

**Review by pacman100 — COMMENTED** [review]
Thank you @SumanthRH for adding IA3 (Infused Adapter by Inhibiting and Amplifying Inner Activations)  PEFT method 🤗🚀✨! This will provide more options for the community to explore for their specific problems. 

Could you please remove changes from existing example notebooks and only have new notebook examples specific to IA3?

I left a comment regarding refactoring some code.

**Comment by pacman100** [comment]
Also, please run `make style` and `make quality` to resolve the quality issues

**Comment by pacman100** [comment]
Hello, it would be great if you could add a test file for testing the minimal core components only (forward, save_pretrained, generate). cc @younesbelkada

**Comment by SumanthRH** [comment]
Hi @pacman100, 

Thanks for checking out the pull request. I can add tests similar to what you have for LoRA/Prefix tuning. I wasn't sure if you would be merging this, so I hadn't added those changes yet.

File: src/peft/utils/save_and_load.py
Comment (SumanthRH) [reply src/peft/utils/save_and_load.py]: Hi. I've made it a bit cleaner now. Please take a look.

**Comment by SumanthRH** [comment]
Hi @pacman100 @younesbelkada 

I looked through the `tests` folder and thought it was better to add IA3 to the existing set of common tests in `tests_common`. Currently, here are all the tests that are supported:

- `_test_model_attr` 
- `_test_prepare_for_training`
- `_test_save_pretrained`
- `_test_merge_layers` (Caveat here: With LoRA, the initialization is random - thus the results after and before merging differ slightly. With IA3, the initialization of the learned vectors is all ones - as per the paper. So the results don't change. Not sure if I should do any additional tests here?)
- `_test_generate`
- `_test_training`

All of these tests are run for decoder models (T5 and BART) and the encoder-decoder models (GPT2, OPT, BLOOM, GPT-NeoX, GPT-Neo). Also, the original authors only specific the IA3 idea for the T0 model, so for other architectures I've simply made an appropriate choice (there could be other choices that are better for that architecture?). Let me know if this works! 

TODO:

- [ ] Half-precision support 
- [ ] Gradient checkpointing
- [ ] Run training with GPT-2 and compare performance (additional sanity check for the `fan_in_fan_out` implementation )

**Comment by SumanthRH** [comment]
Hi @pacman100, @younesbelkada 

I've added half-precision support and a test for gradient checkpointing as well. Please have a look at the current code. Specific attention to half-precision implementation and the implementation for the `modules_to_save` option in `ia3.py` would be helpful (I'm not too sure of this config parameter).  The `examples`  folder has two notebooks - one for sequence classification, and another for seq2seq. 

Some comments on performance:  I believe the original authors tested (IA)^3 with LoRA rank 4, and they found that (IA)^3 outperforms LoRA overall. I did some (perhaps scrappy) benchmarking with rank 8 (default used in peft) and found that with other architectures, especially ones with other trainable parameters like a trainable classification head, (IA)^3 is able to come close to the performance of LoRA, but LoRA is often a bit better. Also, similar to prefix tuning, (IA)^3 takes higher learning rates -  and might need more epochs to converge.

**Comment by SumanthRH** [comment]
Hello @pacman100 , @younesbelkada . I cleaned up the code a bit more. The example notebooks might still need some edits as there's some code that's specific to my username (model saving part). But apart from that, I hope this looks good!

**Review by BenjaminBossan — COMMENTED** [review]
Thanks a lot @SumanthRH for this addition. It looks really nice for the most part and is consistent with the existing code base, well done.

I'm not an expert on this topic (yet), so I cannot comment on every detail. I made a few suggestions, please take a look if they make sense.

In general, I think a docs entry would be good to have, since this introduces a completely new method to the library.

A question more in the direction of the other maintainers: There is a lot of code duplication between this and LoRA. Are we fine with keeping it as or would it be better to add some abstraction? As always, there are pros and cons to both. I just want to ensure that this duplication is indeed the desired state.

File: src/peft/tuners/ia3.py
Comment (SumanthRH) [reply src/peft/tuners/ia3.py:213]: Yes! This works for any Conv1D layer. In fact, this part of the code is borrowed from the LoRA implementation.

File: src/peft/tuners/ia3.py
Comment (SumanthRH) [reply src/peft/tuners/ia3.py]: Done!

File: src/peft/tuners/ia3.py
Comment (SumanthRH) [reply src/peft/tuners/ia3.py]: This should be looking much better now! I've refactored this function based on the new changes from lora

Comment (BenjaminBossan) [reply src/peft/tuners/ia3.py]: Great, the code is indeed more readable now, thanks for making the change.

File: src/peft/tuners/ia3.py
Comment (BenjaminBossan) [reply src/peft/tuners/ia3.py]: Good point about LoRA already doing the same. I was a bit confused because, for instance, [this post](https://stackoverflow.com/a/56685503) implies that Conv1D is only equivalent for kernel size of 1.

Comment (SumanthRH) [reply src/peft/tuners/ia3.py]: Oh my bad, I understood your question now. You're completely right, this is not for _any_ Conv1D layer, in terms of the general 1D convolution layer. This is only for instances of Conv1D in `transformers.pytorch_utils` , as implemented here: https://github.com/huggingface/transformers/blob/9a5d468ba0562e2d5edf9da787881fa227132bca/src/transformers/pytorch_utils.py#L85.  This is equivalent and also implemented as a matrix multiply.

 I was thinking only in terms of the usage of this "Conv1D" in `transformers.pytorch_utils`  (which is a `kernel_size`=1,` stride`=1 version of the general Conv1D in PyTorch here: https://github.com/pytorch/pytorch/blob/d0e5c681f5491fc5bceb3a7f18830f7c7437d7d2/torch/nn/modules/conv.py#L180)

**Comment by SumanthRH** [comment]
Also @BenjaminBossan, I noticed your comment on documentation a bit late, sorry about that! I've added some documentation for $\text{IA}^3$ now, with a basic conceptual guide. Docs should get updated [here](https://moon-ci-docs.huggingface.co/docs/peft/pr_578). In terms of supported methods/models, I've only indicated those covered by `tests` or in `examples`. Let me know if that works!

File: src/peft/tuners/ia3.py
Comment (BenjaminBossan) [reply src/peft/tuners/ia3.py]: Oh I see, thanks a lot for the explanation. I automatically assumed that `Conv1D` was from PyTorch, I should have checked first. Now it makes sense to me.

**Comment by BenjaminBossan** [comment]
> I've added some documentation for IA3 now, with a basic conceptual guide

Fantastic, well written, thank you.

File: src/peft/tuners/ia3.py
Comment (younesbelkada) [inline src/peft/tuners/ia3.py:121]: maybe this was not intended?

File: src/peft/tuners/ia3.py
Comment (younesbelkada) [inline src/peft/tuners/ia3.py:116]: `### new code sampel ###` to remove?

File: src/peft/tuners/ia3.py
Comment (younesbelkada) [inline src/peft/tuners/ia3.py:149]: Can you also add a check that 4bit + IA3 is not supported yet ? (the support can be added in a follow up PR)

File: src/peft/tuners/ia3.py
Comment (younesbelkada) [inline src/peft/tuners/ia3.py:325]: Maybe I am wrong but it seems these methods are never used? the maintained method seems to be `merge_and_unload` and users should use this method rather than `merge_adapter` and `unmerge_adapter`. cc @pacman100

File: tests/testing_common.py
Comment (younesbelkada) [inline tests/testing_common.py:311]: can you elaborate on that? what do you mean specifcally by `merge does not change logits for IA3`? if merging is not supported I think we should just raise an error. WDYT?

**Review by younesbelkada — COMMENTED** [review]
Thanks a lot for your great work @SumanthRH , looking forward to merging this! 
I left few open questions, let me know what do you think
Also @BenjaminBossan raised a great question, regarding code de-duplication we should maybe start by doing the same approach as what we have in transformers `# Copied from ..` and test if the method/ class is effectively copied from somewhere else. I would say let's take care of that in a follow up PR

File: src/peft/tuners/ia3.py
Comment (SumanthRH) [reply src/peft/tuners/ia3.py:325]: You're right! These are methods I'd copied over from LoRA, but they seem to be never used in LoRA as well. I'm removing them from IA3 for now.

File: tests/testing_common.py
Comment (SumanthRH) [reply tests/testing_common.py:311]: So merging is supported, but the specific check here is testing for a difference between `logits_merged`  (after merging IA3 vectors) and `logits_transformers` (without IA3 vectors). With IA3, the initialization is all ones initially (From the paper: "...[the learned vectors] are all initialized with ones so that the overall function computed by the model does not change when they are added." In the case of LoRA, the random initialization will change model outputs ever so slightly after adding LoRA weights, which is being tested with the `assertFalse` statement.  This isn't going to be true for IA3 -  so I have simply ignored that test. I'm not sure what other tests can be added though for merging.

File: src/peft/tuners/ia3.py
Comment (SumanthRH) [reply src/peft/tuners/ia3.py]: This was a typo - the line's supposed to have been added for example config usage. It's fixed now.

File: src/peft/tuners/ia3.py
Comment (SumanthRH) [reply src/peft/tuners/ia3.py:152]: How does this look now?

File: src/peft/tuners/ia3.py
Comment (younesbelkada) [reply src/peft/tuners/ia3.py]: Awesome thank you !

File: src/peft/tuners/ia3.py
Comment (younesbelkada) [reply src/peft/tuners/ia3.py:152]: Sounds great !

File: tests/testing_common.py
Comment (younesbelkada) [reply tests/testing_common.py:311]: I see thanks a lot for the detailed explanation ! Do you think it is possible to force the initialization to be something different than all ones.? I think in the past I have faced the same issue with LoRA because the B matrix was initialized to all zeros. I have added a boolean here: https://github.com/huggingface/peft/blob/06fd06a4d2e8ed8c3a253c67d9c3cb23e0f497ad/src/peft/tuners/lora.py#L91 let me know if you need help on that ! 
 Otherwise all good I think, it can be done in a follow up PR

**Review by younesbelkada — APPROVED** [review]
Looking great on my side ! Thansk for all your great work on this ! 
Let's see what @pacman100 & @BenjaminBossan will say, let me know if I can help you addressing last comments and incoming comments
Thanks again!

File: tests/testing_common.py
Comment (SumanthRH) [reply tests/testing_common.py:311]: Ah okay! There's actaully an argument already called `init_ia3_weights` that's supposed to give this control - but even with `init_ia3_weights` as False, currently the initialization is all ones : 
https://github.com/SumanthRH/peft/blob/5ed92fad11aeec3ec94c8dfd908e81a3cb09d4e4/src/peft/tuners/ia3.py#L399

If this going to be different, then the initialization should be such that the merged version is close enough to the pre-trained weights, which means that the entries in the learned vectors should be close enough to one (because they rescale activations). I'm not too sure what standard initialization fits in here. Maybe something like PyTorch's nn.Linear [initialization](https://pytorch.org/docs/stable/generated/torch.nn.Linear.html) for matrices, but shifted by one: $U(-\sqrt{k}, \sqrt{k}) + 1$ , where $k=1/\text{vector-length}$.  I also don't want to make it look like we conjured up an initialization, so I'm open to hearing your thoughts on this!

**Review by BenjaminBossan — APPROVED** [review]
Overall, this LGTM and I don't have much to add besides what Younes said.

It is not quite clear to me why `_check_target_module_exists` differs from the LoRA implementation, but I think we'll refactor some of that soon, so maybe we can keep it as is.

> Also @BenjaminBossan raised a great question, regarding code de-duplication we should maybe start by doing the same approach as what we have in transformers # Copied from .. and test if the method/ class is effectively copied from somewhere else. I would say let's take care of that in a follow up PR

Okay, let's deal with that later. I'm not sure if the initial reason why `# Copied from` was introduced in transformers necessarily apply here as well, but we can discuss it then.

File: tests/testing_common.py
Comment (BenjaminBossan) [reply tests/testing_common.py:311]: @younesbelkada Could you explain why we want to require the outputs to be different each time when initializing the model? Is it not actually a good thing if by default, the added weights lead to an identity operation? I'm running into the same test error when using LoRA with certain custom models.

Comment (pacman100) [reply tests/testing_common.py:311]: Hello @BenjaminBossan, LoRA initializes `lora_B` to all 0 so that we start with an identity function. So, for all usage `init_lora_weights=True`. Only during testing when we can't actually train the model but want to test the functionality of `merge_and_unload`, `init_lora_weights` is set to False else even after merging it will be identity op and we can't test if the merge was successful or not.

**Review by pacman100 — APPROVED** [review]
Thank you so much @SumanthRH for all the efforts put it to add IA3. Left a suggestion, other than that LGTM!

File: tests/testing_common.py
Comment (BenjaminBossan) [reply tests/testing_common.py:311]: Ah I see, thanks for explaining. This should probably added as a comment to that line, since I think I'm not the only one who was confused.

Regarding the testing of the merging feature, would it be possible to check the weights directly, instead of the model outputs?

File: src/peft/tuners/ia3.py
Comment (SumanthRH) [reply src/peft/tuners/ia3.py]: Missed this! Thanks! Have committed your suggestion now!

**Review by pacman100 — APPROVED** [review]

**Comment by SumanthRH** [comment]
Happy to see this merged! A small note to the maintainers @pacman100 @younesbelkada: The example notebooks [peft_ia3_seq2seq.ipynb](https://github.com/huggingface/peft/blob/main/examples/conditional_generation/peft_ia3_seq2seq.ipynb) and [IA3.ipynb](https://github.com/huggingface/peft/blob/main/examples/sequence_classification/IA3.ipynb) might need some minor changes (specifically, in the "Share adapters to the Hub" sections at the end).  Other than that, we should be okay! This was also my first major OSS contribution, so thanks for helping me out!

**Pull request merged**
